### PR TITLE
tests: set memory limit for snapd

### PR DESCRIPTION
**Pull request by bboozzoo**
Make sure that snapd isn't doing anything silly and use a reasonable a memory
limit when running the test suite to catch problems early.

### CODE REVIEW

**Comment by pedronis** [comment]
wouldn't snapd just be killed and restarted if hit this? what would show up in the logs in that case? should we add a check in one of the restores about nothing like that showing up in the logs?

**Review by pedronis — CHANGES_REQUESTED** [review]
this doesn't seem quite all we need, see my comment

**Review by mardy — APPROVED** [review]

**Comment by mardy** [comment]
> wouldn't snapd just be killed and restarted if hit this? what would show up in the logs in that case? should we add a check in one of the restores about nothing like that showing up in the logs?

Unless we have some tests relying on the service being restarted, we could add `Restart=no` to the unit file.

**Review by pedronis — APPROVED** [review]
thank you

**Comment by pedronis** [comment]
@mardy 
> Unless we have some tests relying on the service being restarted, we could add Restart=no to the unit file.

lot's of behavior of snapd relies on the service being restarted, first of all refreshes/upgrades of snapd

File: tests/lib/prepare-restore.sh [hunk 1]
```diff
@@ -717,6 +717,8 @@ restore_project_each() {
     # will most likely not function correctly anymore. It looks like this
     # happens with: https://forum.snapcraft.io/t/4101 and is a source of
     # failure in the autopkgtest environment.
+    # Also catch a scenario when snapd service hits the MeoryMax limit set while
```
Comment (mardy) [inline tests/lib/prepare-restore.sh:720]: MeoryMax -> MemoryMax :-)

**Comment by mardy** [comment]
> @mardy
> 
> > Unless we have some tests relying on the service being restarted, we could add Restart=no to the unit file.
> 
> lots of behavior of snapd relies on the service being restarted, first of all refreshes/upgrades of snapd

Even assuming that it also applies to the test suite, we could use `Restart=on-success`: so, if snapd gets killed by the OOM, it won't get restarted and tests will likely fail.

File: tests/lib/prepare-restore.sh (hunk 1 above)
Comment (bboozzoo) [reply tests/lib/prepare-restore.sh:720]: :facepalm: thanks!

**Comment by mardy** [comment]
> It's complex. And of course the test suite exercises quite a bit of it. Basically it is not a good idea to change that behavior globally in the test suite, because then we are not testing what we actually want to test.

Actually, I now realized that the systemd default is `Restart=no`, so the test suite is already relying on the fact that snapd will never be restarted. :-)

**Comment by pedronis** [comment]
> Actually, I now realized that the systemd default is Restart=no, so the test suite is already relying on the fact that snapd will never be restarted. :-)

unless I'm very confused most of the spread tests run snapd with the snapd service file as intended which has "always" (plus details), otherwise snapd refreshes wouldn't work

**Review by sergiocazzolato — APPROVED** [review]
Thanks for this change

**Review by mvo5 — APPROVED** [review]
Woah, this is nice

**Comment by mvo5** [comment]
This is great but it seems we trigger the failures already, e.g.:
````
2021-05-07T08:57:01.2212017Z ##[error]2021-05-07 08:57:01 Error restoring google:ubuntu-core-20-64:tests/main/core16-base (may070835-707605) : 
2021-05-07T08:57:01.2284428Z -----
...
2021-05-07T08:57:01.2584218Z + grep oom-killer
2021-05-07T08:57:01.2584769Z + dmesg
2021-05-07T08:57:01.2586867Z [  134.454207] snapd invoked oom-killer: gfp_mask=0x101cca(GFP_HIGHUSER_MOVABLE|__GFP_WRITE), order=0, oom_score_adj=-900
2021-05-07T08:57:01.2588702Z + echo 'oom-killer got invoked during the tests, this should not happen.'
2021-05-07T08:57:01.2590194Z oom-killer got invoked during the tests, this should not happen.
2021-05-07T08:57:01.2591410Z + echo 'Dmesg debug output:'
```
and probably a bunch more of those.

**Comment by bboozzoo** [comment]
Now that we have a better view, I think I can add something to the tests that would temporarily increase the memory limit during a particular test, such that we can isolate those that require more memory but still keep the check on the ones that don't. Preferrably through some environment variable that could be picked up during prepare and then restore the previous bit in the restore phase.

**Comment by bboozzoo** [comment]
The PR has changed quite a bit, so I re-requested reviews. With the latest update, snapd runs with a memory limit on its cgroup, unless SNAPD_NO_MEMORY_LIMIT environment variable is set in which case there's no limit at all.

File: tests/lib/prepare.sh [hunk 2]
```diff
@@ -193,6 +193,24 @@ update_core_snap_for_classic_reexec() {
     done
 }
 
+prepare_memory_limit_override() {
+    if [ -n "${SNAPD_NO_MEMORY_LIMIT-}" ]; then
+        return
+    fi
+    # Use MemoryMax to set the memory limit for snapd.service, that is the main
+    # snapd snapd process and its subprocesses executing within the same cgroup.
```
Comment (mardy) [inline tests/lib/prepare.sh:201]: nitpick: there's one "snapd" too many in this line.

File: tests/lib/state.sh [hunk 3]
```diff
@@ -90,6 +90,9 @@ restore_snapd_state() {
         restore_snapd_lib
         cp -rf "$SNAPD_STATE_PATH"/snapd-cache/*  /var/cache/snapd
         cp -rf "$SNAPD_STATE_PATH"/boot/* "$boot_path"
+        # Purge all the systemd service units config
```
Comment (mardy) [inline tests/lib/state.sh:93]: nitpick: at the first sight, it looks like we are removing the unit files themselves. Adding a "overrides" at the end of the comment would make it more clear.

**Review by mardy — APPROVED** [review]
Just nitpicks, it's also very fine as it is :-)

File: tests/lib/prepare.sh [hunk 4]
```diff
@@ -193,6 +193,25 @@ update_core_snap_for_classic_reexec() {
     done
 }
 
+prepare_memory_limit_override() {
+    if [ -n "${SNAPD_NO_MEMORY_LIMIT-}" ]; then
```
Comment (anonymouse64) [inline tests/lib/prepare.sh:197]:
```suggestion
    if [ -n "${SNAPD_NO_MEMORY_LIMIT:-}" ]; then
```

**Review by anonymouse64 — APPROVED** [review]
one typo, but otherwise looks good to me

**Review by sergiocazzolato — APPROVED** [review]
Thanks for the change, I would like to test this in all the supported devices once it is merged.

**Comment by sergiocazzolato** [comment]
I see in some tests whith SNAPD_NO_MEMORY_LIMIT: 1 some broken pipe errors like

+ snap install --dangerous /var/lib/snapd/snaps/core_11734.snap
error: cannot communicate with server: Post http://localhost/v2/snaps: write unix @->/run/snapd.socket: write: broken pipe
-----
Errors in xenial:

2021-08-27 16:26:44 Failed tasks: 5
    - google:ubuntu-16.04-64:tests/main/disk-space-awareness
    - google:ubuntu-16.04-64:tests/main/snap-userd-reexec
    - google:ubuntu-16.04-64:tests/main/snapd-certs
    - google:ubuntu-16.04-64:tests/main/snapd-reexec
    - google:ubuntu-16.04-64:tests/upgrade/basic

**Comment by sergiocazzolato** [comment]
@anonymouse64  by mistake I removed the approve that you did

File: tests/lib/prepare.sh [hunk 5]
```diff
@@ -192,6 +192,54 @@ update_core_snap_for_classic_reexec() {
...
+    local set_limit=1
+
+    case "$SPREAD_SYSTEM" in
+        ubuntu-core-16-*|ubuntu-core-18-*|ubuntu-16.04-*)
+            # the tests on UC16 and UC18 have demonstrated that the memory limit
+            # state claimed by systemd may be out of sync with actual memory
+            # controller setting for the snapd.service cgroup
+            set_limit=0
+            ;;
+        amazon-linux-*)
```
Comment (anonymouse64) [inline tests/lib/prepare.sh:207]: I would have guessed Centos 7 is in the same boat as amazon-linux ?

File: tests/lib/prepare.sh [hunk 6]
```diff
@@ -192,6 +192,54 @@ update_core_snap_for_classic_reexec() {
...
+        # cgroup. If snapd hits the memory limit, it will get killed by
+        # oom-killer which will be caught in restore_project_each in
+        # prepare-restore.sh.
+        #
+        # This ought to set MemoryMax, but on systems with older systemd we need to
+        # use MemoryLimit, which is deprecated and replaced by MemoryMax now, but
+        # systemd is backwards compatible so the limit is still set.
+        cat <<EOF > /etc/systemd/system/snapd.service.d/memory-max.conf
+[Service]
+MemoryLimit=100M
```
Comment (anonymouse64) [inline tests/lib/prepare.sh:234]: why not set both like we do for quota groups?

**Review by anonymouse64 — APPROVED** [review]
lgtm but a question

File: tests/lib/prepare.sh (hunk 5 above)
Comment (bboozzoo) [reply tests/lib/prepare.sh:222]: actually the tests did not fail on centos 7, but I had to workaround some weird inconsistency on 16.04 and 18.04. Perhaps those are real bugs in systemd that got patched in rhel.

File: tests/lib/prepare.sh (hunk 6 above)
Comment (bboozzoo) [reply tests/lib/prepare.sh:249]: Missed this one. I'm waiting for spread run to complete. If it lands soonish I'll open a followup.

**Comment by bboozzoo** [comment]
@mvo5 this branch is ready, test failures are unrelated, can you merge it?

**Pull request merged**